src/main/java/com/shaft/support/DBActions.java

```diff
@@ -0,0 +1,155 @@
+package com.shaft.support;
+
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+
+import org.testng.Assert;
+
+import com.shaft.io.ReportManager;
+
+public class DBActions {
+
+	Connection connection;
+	Statement statement;
+	ResultSet resultSet;
+	String dbType;
+	String dbServerIP;
+	String dbPort;
+	String dbName;
+	String username;
+	String password;
+
+	public DBActions(String dbType, String dbServerIP, String dbPort, String dbName, String username, String password) {
+		this.dbType = dbType;
+		this.dbServerIP = dbServerIP;
+		this.dbPort = dbPort;
+		this.dbName = dbName;
+		this.username = username;
+		this.password = password;
+	}
+
+	private void passAction(String actionName, String testData, String log) {
+		String message = "Successfully performed action [" + actionName + "].";
+		if (testData != null) {
+			message = message + " With the following test data [" + testData + "].";
+		}
+		ReportManager.log(message);
+		if (log != null) {
+			ReportManager.attachAsStep("Passed", "DB Connection Log", log);
+		}
+	}
+
+	private void passAction(String actionName, String testData) {
+		passAction(actionName, testData, null);
+	}
+
+	private void failAction(String actionName, String testData, String log) {
+		String message = "Failed to perform action [" + actionName + "].";
+		if (testData != null) {
+			message = message + " With the following test data [" + testData + "].";
+		}
+		ReportManager.log(message);
+		if (log != null) {
+			ReportManager.attachAsStep("Failed", "DB Connection Log", log);
+		}
+		Assert.fail(message);
+	}
+
+	private void failAction(String actionName, String testData) {
+		failAction(actionName, testData, null);
+	}
+
+	/**
+	 * Setup connection to database
+	 * 
+	 * @throws ClassNotFoundException
+	 * @throws SQLException
+	 */
+	public void createConnection() {
+		String testData = dbServerIP + ":" + dbPort + "/" + dbName;
+
+		try {
+			switch (dbType) {
+			// create connection to mysql DB
+			case ("mysql"):
+				Class.forName("com.mysql.jdbc.Driver");
```

Comment:
I understand that this line can sometimes be used as a workaround for some broken java implementation, and that sometimes it has a .newInstance(); at the end of it. Please explain this case via in-code comments as it currently looks like an unneeded call for a class that is never used.

Reply:
I don't understand what is the workaround for some broken java implementation. could you explain more please?

this line is used to specify to DriverManager which JDBC drivers to try to make Connections with.

Reply:
This line is not mandatory, as per the below docs:
https://docs.oracle.com/javase/8/docs/api/java/sql/DriverManager.html#getConnection-java.lang.String-

Instead, you should use the register method, and then simply get connection.

Here is a sample for this discussion:
https://stackoverflow.com/questions/5484227/jdbc-class-forname-vs-drivermanager-registerdriver

This line is used to load a class that was not loaded, and this is a workaround in case the regular register / connect approach fails for some reason. But it's not the best practice.

I assumed that you faced this issue and therefor used this workaround, and therefor asked that you document the issue that you faced.

In case there was no issue using the regular register then get connection, please refactor the code accordingly.

Reply:
Done, as per first link, I removed the method and I didn't face any issue:

Class.forName() is not directly related to JDBC at all. It simply loads a class.
Most JDBC Driver classes register themselves in their static initializers by calling registerDriver().
registerDriver() is the real call that you hardly ever need to call yourself (unless you write your own JDBC driver).

Reply:
(Y)